File: crates/fieldglass/tests/conformance.rs
//! The reference runner for the ADR-0006 conformance suite (#573).
//!
//! Three runners drive `conformance/suite.json`. This is the first and the
//! one that records it:
//!
//! | runner | binding under test | where |
//! |---|---|---|
//! | this file | `fieldglass::Session` itself | `cargo test --workspace`, and `cargo test --target wasm32-wasip1 -p fieldglass` in CI |
//! | `fieldglass-napi`'s `conformance_host` module | the napi handles | `cargo test --workspace` |
//! | `crates/fieldglass-wasm/tests/node/conformance.mjs` | the built browser bundle, from Node | the `wasm32 build` CI job |
//!
//! The first two are Rust and the third is JavaScript, which is the point:
//! ADR-0006 decision 3 asks for expectations that live in the crate as *data*,
//! so a host in another language reads the same file rather than a translated
//! copy of it.
//!
//! # Recording
//!
//! ```sh
//! FIELDGLASS_UPDATE_CONFORMANCE=1 cargo test -p fieldglass --test conformance
//! ```
//!
//! The run fails afterwards on purpose: a run that re-recorded has verified
//! nothing, and an environment with the variable left set must not be able to
//! turn the suite into a no-op that re-baselines every diff it exists to catch.

use fieldglass::conformance::{
    self, Case, RecordedCase, Suite, Tolerance, cases, compare, error_codes, observe,
};

/// Set to re-record rather than compare.
const UPDATE_ENV: &str = "FIELDGLASS_UPDATE_CONFORMANCE";

/// The recording, relative to this crate's manifest directory (which cargo
/// makes the working directory of a test).
const SUITE_PATH: &str = "conformance/suite.json";

/// Read a fixture named the way the suite names one: relative to `crates/`.
fn read(fixture: &str) -> Result<Vec<u8>, String> {
    let path = format!("../{fixture}");
    std::fs::read(&path).map_err(|e| format!("{path}: {e}"))
}

/// The shipped suite, parsed, with a message that says what to do when it is
/// the file rather than the code that is wrong.
fn shipped() -> Suite {
    conformance::suite().unwrap_or_else(|e| {
        panic!("{SUITE_PATH} does not parse as the current `Suite` type: {e}\nre-record with {UPDATE_ENV}=1")
    })
}

/// Rewrite the recording from a fresh run, then fail.
fn rerecord() {
    let suite = conformance::record(&read).expect("every fixture reads");
    let json = conformance::to_pretty_json(&suite).expect("the suite serialises");
    std::fs::write(SUITE_PATH, json).expect("the recording is writable");
    panic!(
        "{SUITE_PATH} rewritten from this run ({} cases). \
         Unset {UPDATE_ENV} and run again to verify it.",
        suite.cases.len()
    );
}

/// The suite, case by case, against `fieldglass::Session`.
///
/// This is the recording run as well: with `FIELDGLASS_UPDATE_CONFORMANCE` set
/// it rewrites the file and then fails, so re-recording can never be mistaken
/// for a green run.
#[test]
fn every_conformance_case_matches_its_recording() {
    if std::env::var_os(UPDATE_ENV).is_some() {
        rerecord();
    }
    let suite = shipped();
    assert!(
        !suite.cases.is_empty(),
        "the conformance suite is empty — it would pass while checking nothing"
    );

    let mut failures = Vec::new();
    for RecordedCase { case, expect } in &suite.cases {
        let bytes = read(&case.fixture).expect("fixture");
        let observed = observe(&bytes, case);
        for line in compare(expect, &observed, suite.tolerance) {
            failures.push(format!("{}: {line}", case.id));
        }
    }
    assert!(
        failures.is_empty(),
        "{} conformance disagreement(s):\n{}",
        failures.len(),
        failures.join("\n")
    );
}

/// The file must still describe exactly the cases the code names, in order.
///
/// Without this, adding a case and forgetting to re-record would leave the new
/// case simply absent — the suite would stay green while covering less, which
/// is the failure mode a data-driven gate is most prone to.
#[test]
fn the_recording_covers_exactly_the_cases_the_code_names() {
    let suite = shipped();
    let recorded: Vec<Case> = suite.cases.iter().map(|r| r.case.clone()).collect();
    let named = cases();
    assert_eq!(
        recorded.len(),
        named.len(),
        "the recording holds {} cases and the code names {} — re-record with {UPDATE_ENV}=1",
        recorded.len(),
        named.len()
    );
    for (r, n) in recorded.iter().zip(&named) {
        assert_eq!(r, n, "recorded case differs from the one the code names");
    }
}

/// Every code in the shipped list is produced by a real call in the suite, and
/// every code a suite case produced is in the list.
///
/// This is what "`Error::code()` values are pinned by the suite" means in
/// practice. A list on its own would pass while naming codes nothing can
/// reach; the cases on their own would pass while a variant quietly stopped
/// being reachable at all.
#[test]
fn every_error_code_is_both_listed_and_reachable() {
    let suite = shipped();
    assert_eq!(
        suite.error_codes,
        error_codes(),
        "the recorded error codes differ from this build's — re-record with {UPDATE_ENV}=1"
    );

    let mut seen: Vec<String> = suite
        .cases
        .iter()
        .filter_map(|r| {
            r.expect
                .get("error")
                .and_then(|e| e.get("code"))
                .and_then(|c| c.as_str())
                .map(str::to_string)
        })
        .collect();
    seen.sort();
    seen.dedup();
    assert_eq!(
        seen,
        error_codes(),
        "every error code must be produced by at least one conformance case"
    );
}

/// Exactly these cases record a failure; every other one produced an answer.
///
/// The five `error/…` cases exist to prove each [`fieldglass::Error`] code is
/// reachable. The two beside them are the `degenerate` subject — a §3.20 grid
/// stating `Dx = Dy = 0`, which decodes fine and has no extent to warp onto,
/// which is why that fixture is in the suite. `degenerate/warp/window` is not
/// here: a *manual* window gives the warp a box even when the grid states none,
/// and that difference is worth having recorded.
///
/// `error/unsupported` is that same refusal, on purpose: since #580 no fixture
/// in the corpus produces `unsupported` at `decode`, because both families that
/// used to — spectral and HEALPix — are synthesised onto a lat/lon grid now.
/// It is kept as a named case rather than left to `degenerate/warp/bilinear`
/// so that the code's reachability is proved by a case whose whole job that is.
const CASES_THAT_RECORD_A_FAILURE: &[&str] = &[
    "degenerate/warp/bilinear",
    "degenerate/warp/nearest",
    "error/decode",
    "error/invalid_option",
    "error/no_such_message",
    "error/unsupported",
    "error/unsupported_format",
];

/// A case that stopped exercising its operation records a failure instead, and
/// passes for ever after.
///
/// This is the hole in a recorded suite: [`observe`] turns *any* failure into an
/// observation, so an argument dropped in a refactor, a fixture that no longer
/// decodes, or an option an operation newly rejects re-records as
/// `{"error": …}` and nothing says so. The recording still matches, the case
/// count is still right, and `every_error_code_is_both_listed_and_reachable` is
/// satisfied by the deliberate cases whatever else joins them.
///
/// Asserted in both directions, because half of it would be the same fail-open:
/// a case that starts failing has to be added here on purpose, and one that
/// stops failing — a refusal quietly becoming an answer — is a change to the
/// contract too.
#[test]
fn exactly_the_expected_cases_record_a_failure() {
    let suite = shipped();
    let mut failing: Vec<&str> = suite
        .cases
        .iter()
        .filter(|r| r.expect.get("error").is_some())
        .map(|r| r.case.id.as_str())
        .collect();
    failing.sort_unstable();
    assert_eq!(
        failing, CASES_THAT_RECORD_A_FAILURE,
        "a case that records a failure proves nothing about the operation it \
         names; add it here deliberately or find out why it stopped answering"
    );
}

/// The recorded tolerance is the one this build chose.
///
/// It is read out of the JSON and used by all three runners, so editing that
/// one line to `1e-1` would loosen every comparison in the suite at once and
/// leave nothing red — including the perturbation check below, which is
/// supposed to be the thing that proves the numbers are pinned.
#[test]
fn the_recorded_tolerance_is_the_one_this_build_chose() {
    assert_eq!(
        shipped().tolerance,
        Tolerance::default(),
        "the suite's tolerance has been edited away from this build's"
    );
}

/// The comparator rejects a recording perturbed by more than its tolerance —
/// for every case that has a real-valued leaf to perturb.
///
/// Without this the suite could be green because the comparison is loose rather
/// than because the answers agree.
#[test]
fn the_suite_would_notice_if_the_answers_moved() {
    let suite = shipped();
    // The suite's own tolerance, not this build's default: a recording that
    // widened it has to fail *here*, which is what the test above pins.
    let tol = suite.tolerance;

    // Perturb every real-valued leaf of every recorded observation by a
    // thousand times the tolerance and check each case notices.
    let mut unmoved = Vec::new();
    for RecordedCase { case, expect } in &suite.cases {
        let mut perturbed = expect.clone();
        let moved = perturb(&mut perturbed);
        if !moved {
            // A case whose observation is entirely discrete (an error code, a
            // count) is pinned by exact comparison already; nothing to check.
            continue;
        }
        if compare(expect, &perturbed, tol).is_empty() {
            unmoved.push(case.id.clone());
        }
    }
    assert!(
        unmoved.is_empty(),
        "these cases accept a perturbed recording, so their numbers are not pinned: {unmoved:?}"
    );
}

/// Move every real-valued leaf well outside the tolerance. Returns whether
/// anything moved.
fn perturb(value: &mut serde_json::Value) -> bool {
    match value {
        serde_json::Value::Number(n) if n.is_f64() => {
            let Some(x) = n.as_f64() else { return false };
            let bumped = x + 1e-6 + x.abs() * 1e-6;
            match serde_json::Number::from_f64(bumped) {
                Some(m) => {
                    *n = m;
                    true
                }
                None => false,
            }
        }
        // A plain loop rather than `any`, which short-circuits: every leaf has
        // to move, or the check would only prove the *first* number is pinned.
        serde_json::Value::Array(a) => {
            let mut moved = false;
            for v in a {
                moved |= perturb(v);
            }
            moved
        }
        serde_json::Value::Object(o) => {
            let mut moved = false;
            for (_, v) in o {
                moved |= perturb(v);
            }
            moved
        }
        _ => false,
    }
}

File: crates/fieldglass-wasm/tests/node/conformance.mjs
// The browser host, run through the ADR-0006 conformance suite (#573).
//
//   crates/fieldglass-wasm/build.sh nodejs
//   node crates/fieldglass-wasm/tests/node/conformance.mjs
//
// The suite is `crates/fieldglass/conformance/suite.json`: cases and expected
// observations, recorded by `fieldglass`'s own runner. Nothing here is a
// translation of that runner — this file reads the same JSON and drives the
// **built wasm module through its JavaScript surface**, which is the only way
// the browser binding's own work (serde-wasm-bindgen conversion, the typed
// arrays, the error mapping) is under test at all.
//
// Three runners share this one file of expectations:
//
//   fieldglass/tests/conformance.rs        Session, native and wasm32-wasip1
//   fieldglass-napi/src/conformance_host   the napi handles
//   this file                              the browser bundle, from Node
//
// It is a CI gate, not a manual check: unlike `parity.mjs` it needs no sample
// corpus, only the committed fixtures and a built `pkg/nodejs`. It therefore
// fails loudly when the bundle is missing rather than skipping — a conformance
// runner that quietly compared nothing is worse than none.
//
// # One difference from the Rust comparator, stated
//
// The Rust side compares JSON integers exactly and only real-valued leaves
// within the tolerance (ADR-0009: the discrete decisions agree exactly across
// targets, so comparing them loosely gives away coverage). JavaScript has one
// number type and `JSON.parse` cannot tell `1` from `1.0`, so this runner
// applies the tolerance to every number. The distinction costs nothing here:
// the tolerance is `1e-9 + 1e-9 * |expected|` and no count, length or byte in
// the suite exceeds ~10^5, so the slack on a discrete value stays below 10^-4
// where the smallest real difference is 1.

import { createRequire } from 'node:module';
import { readFileSync, existsSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const require = createRequire(import.meta.url);
const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = join(here, '..', '..', '..', '..');

// `--pkg <dir>` picks a different build — the `+simd128` one, say, which is
// where a float result would most plausibly move.
const pkgArg = process.argv.indexOf('--pkg');
const pkgDir = pkgArg === -1 ? 'crates/fieldglass-wasm/pkg/nodejs' : process.argv[pkgArg + 1];
const WASM = join(repoRoot, pkgDir, 'fieldglass_wasm.js');
const SUITE = join(repoRoot, 'crates/fieldglass/conformance/suite.json');

for (const [path, how] of [
  [WASM, 'crates/fieldglass-wasm/build.sh nodejs'],
  [SUITE, 'it is committed; check out the repository'],
]) {
  if (!existsSync(path)) {
    console.error(`missing: ${path}\n  ${how}`);
    process.exit(1);
  }
}

const wasm = require(WASM);
const suite = JSON.parse(readFileSync(SUITE, 'utf8'));

// ---------------------------------------------------------------------------
// The observation
// ---------------------------------------------------------------------------

/** `fieldglass::conformance::sample_indices`, to the integer. */
function sampleIndices(len) {
  if (len === 0) return [];
  const out = [0, Math.floor(len / 4), Math.floor(len / 2), Math.floor(len / 4) * 3, len - 1];
  return [...new Set(out)].sort((a, b) => a - b);
}

/** A real-valued leaf, in the spelling `fieldglass::conformance::real` uses:
 *  `null` for absent, the number when it is finite, and the *name*
 *  `"nonFinite"` otherwise. Deliberately not `"NaN"` / `"Infinity"`: JavaScript
 *  and Rust spell those differently, and which non-finite it was is not
 *  something the hosts have to agree on. */
function real(v) {
  if (v === null || v === undefined) return null;
  return Number.isFinite(v) ? v : 'nonFinite';
}

/** `undefined` is how napi and serde-wasm-bindgen both spell Rust's `None`;
 *  the recording spells it `null`. Normalise so a missing key and an absent
 *  value stay distinguishable from each other.
 *
 *  Only plain objects and arrays are walked. A `Map` — what serde-wasm-bindgen
 *  gives a Rust map by default — comes back untouched, and `compare` then
 *  reports every key as missing, which is a loud failure rather than a silent
 *  pass. No DTO holds a map today; if one does, this is where to teach it. */
function nulled(v) {
  if (v === undefined) return null;
  if (Array.isArray(v)) return v.map(nulled);
  if (v && typeof v === 'object' && v.constructor === Object) {
    return Object.fromEntries(Object.entries(v).map(([k, x]) => [k, nulled(x)]));
  }
  return v;
}

/** A `Georef` as the suite records one: everything but `geometry`, which is
 *  `core`'s tagged enum and deliberately outside the host contract. */
function georef(grid) {
  const { geometry, ...rest } = nulled(grid);
  void geometry;
  return rest;
}

function paletteOptions(args) {
  return {
    colormap: args.colormap ?? null,
    reversed: args.reversed ?? false,
    min: args.rangeMin ?? null,
    max: args.rangeMax ?? null,
    scale: args.scale ?? null,
  };
}

function decodeOptions(args) {
  return { dtype: args.dtype ?? 'auto' };
}

/** What a decoded field contributes to an observation —
 *  `fieldglass::conformance::field_value`. Shared by `decode` and `combine`,
 *  which record the same shape because `combine` answers a `Field` too. */
function fieldObservation(field) {
  const values = field.values();
  const mask = field.mask();
  return {
    dtype: field.dtype(),
    len: values.length,
    maskLen: mask.length,
    maskOnes: mask.reduce((n, m) => n + (m === 1 ? 1 : 0), 0),
    ni: field.ni(),
    nj: field.nj(),
    parameter: field.parameter(),
    units: field.units(),
    stats: nulled(field.stats()),
    georef: georef(field.grid()),
    samples: sampleIndices(values.length).map((i) => ({
      i,
      // Read the mask first, as the DTO's own doc says: the buffer holds
      // *something* at a masked slot and it is not data.
      v: mask[i] === 1 ? real(values[i]) : null,
      m: mask[i],
    })),
  };
}

/** What the browser host answers for one case. Failures become the same
 *  `{ error: { code, hasMessage } }` shape the Rust runner records — and the
 *  code is really there, because `throw()` sets it on the JS `Error`. */
function observe(caseSpec) {
  const path = join(repoRoot, 'crates', caseSpec.fixture);
  let bytes = readFileSync(path);
  if (caseSpec.args.truncate !== null && caseSpec.args.truncate !== undefined) {
    bytes = bytes.subarray(0, Math.min(caseSpec.args.truncate, bytes.length));
  }
  try {
    return run(caseSpec, new Uint8Array(bytes));
  } catch (e) {
    return {
      error: {
        code: e && typeof e.code === 'string' ? e.code : null,
        hasMessage: Boolean(e && e.message),
      },
    };
  }
}

function run(caseSpec, bytes) {
  const handle = wasm.open(bytes);
  // The host's own memory contract: linear memory never shrinks, so both the
  // field and the handle are released as soon as the observation is built.
  // 149 leaked sessions over the committed corpus is not much, but a runner
  // that ignores the contract it is checking is not a good example of it.
  try {
    return withHandle(caseSpec, handle);
  } finally {
    handle.free();
  }
}

function withHandle(caseSpec, handle) {
  const { op, args } = caseSpec;

  if (op === 'open') {
    return { format: handle.format(), count: handle.count(), addressing: handle.addressing() };
  }
  if (op === 'message') {
    const info = nulled(handle.message(args.index));
    if (info.grid) info.grid = georef(info.grid);
    return info;
  }
  // The variable addressing mode (#679): a listing, and a slice of a variable
  // rather than a message decoded by index.
  if (op === 'variables') return nulled(handle.variables());
  if (op === 'dimensions') return nulled(handle.dimensions());
  if (op === 'decodeSlice') {
    const slice = handle.decodeSlice(
      args.variable,
      args.yDim,
      args.xDim,
      new Uint32Array(args.sliceIndices ?? []),
      decodeOptions(args),
    );
    try {
      return fieldObservation(slice);
    } finally {
      slice.free();
    }
  }

  // Everything else decodes first. The field is owned by this side, so it is
  // freed as soon as the observation is built.
  const field = handle.decode(args.index, decodeOptions(args));
  try {
    switch (op) {
      case 'decode':
        return fieldObservation(field);
      case 'combine': {
        // Field B is a second decode of the same message: `combine` takes two
        // fields, and every GRIB fixture in the corpus holds one message, so
        // the suite has no second index to point at. Freed here rather than in
        // the `finally` below, which owns field A.
        const b = handle.decode(args.index, decodeOptions(args));
        try {
          const out = handle.combine(field, b, args.combineOp);
          try {
            return fieldObservation(out);
          } finally {
            out.free();
          }
        } finally {
          b.free();
        }
      }
      case 'warp': {
        const out = handle.warp(field, {
          bilinear: args.bilinear ?? true,
          bounds: args.bounds ?? null,
          // The caller's own output raster (#465). Passed through as `null`
          // rather than omitted when the case does not name one, so this
          // adapter states the whole option set the façade takes.
          width: args.width ?? null,
          height: args.height ?? null,
        });
        return {
          width: out.width,
          height: out.height,
          bounds: Array.from(out.bounds),
          len: out.values.length,
          maskOnes: out.mask.reduce((n, m) => n + (m === 1 ? 1 : 0), 0),
          samples: sampleIndices(out.values.length).map((i) => ({
            i,
            v: out.mask[i] === 1 ? real(out.values[i]) : null,
            m: out.mask[i],
          })),
        };
      }
      case 'palette': {
        const p = handle.palette(field, paletteOptions(args));
        return {
          t0: real(p.t0),
          t1: real(p.t1),
          scale: p.scale,
          lutLen: p.lut.length,
          lutSamples: sampleIndices(p.lut.length).map((i) => ({ i, b: p.lut[i] })),
          maskedRgba: Array.from(p.maskedRgba),
        };
      }
      case 'render': {
        const rgba = handle.render(field, paletteOptions(args), args.flipY ?? false);
        let opaque = 0;
        for (let k = 3; k < rgba.length; k += 4) if (rgba[k] === 255) opaque += 1;
        return {
          width: field.ni(),
          height: field.nj(),
          rgbaLen: rgba.length,
          opaque,
          pixels: sampleIndices(rgba.length / 4).map((i) => ({
            i,
            rgba: Array.from(rgba.subarray(i * 4, i * 4 + 4)),
          })),
        };
      }
      case 'probe': {
        const p = handle.probe(field, args.lat ?? 0, args.lon ?? 0);
        return p === undefined ? null : nulled(p);
      }
      case 'contours': {
        const lines = handle.contours(field, new Float64Array(args.levels ?? []));
        return {
          levelCount: lines.length,
          levels: lines.map((l) => real(l.value)),
          segmentCounts: lines.map((l) => l.segments.length),
        };
      }
      default:
        throw new Error(`the suite names an operation this runner does not know: ${op}`);
    }
  } finally {
    field.free();
  }
}

// ---------------------------------------------------------------------------
// The comparison — `fieldglass::conformance::compare`, in JavaScript
// ---------------------------------------------------------------------------

function compare(expected, observed, tol, path = '$', out = []) {
  const kind = (v) => (v === null ? 'null' : Array.isArray(v) ? 'array' : typeof v);
  const ke = kind(expected);
  const ko = kind(observed);
  if (ke !== ko) {
    out.push(`${path}: expected ${ke} ${JSON.stringify(expected)}, got ${ko} ${JSON.stringify(observed)}`);
    return out;
  }
  if (ke === 'object') {
    const a = Object.keys(expected).sort();
    const b = Object.keys(observed).sort();
    const missing = a.filter((k) => !b.includes(k));
    const extra = b.filter((k) => !a.includes(k));
    if (missing.length) out.push(`${path}: missing keys ${JSON.stringify(missing)}`);
    if (extra.length) out.push(`${path}: unexpected keys ${JSON.stringify(extra)}`);
    for (const k of a) if (b.includes(k)) compare(expected[k], observed[k], tol, `${path}.${k}`, out);
    return out;
  }
  if (ke === 'array') {
    if (expected.length !== observed.length) {
      out.push(`${path}: length ${expected.length} != ${observed.length}`);
      return out;
    }
    expected.forEach((v, i) => compare(v, observed[i], tol, `${path}[${i}]`, out));
    return out;
  }
  if (ke === 'number') {
    // Written as a negated `<=`, not as `>`, for one reason: `NaN > x` is
    // false and `NaN <= x` is false, so the `>` spelling accepts a NaN
    // observation against any expectation. The recording can never hold one
    // (`real()` on the Rust side names a non-finite rather than writing it),
    // so a NaN here is always a browser-host result the reference does not
    // share — exactly what this runner exists to catch. The Rust comparator
    // is the same way round.
    if (!(Math.abs(expected - observed) <= tol.absolute + tol.relative * Math.abs(expected))) {
      out.push(`${path}: ${expected} != ${observed}`);
    }
    return out;
  }
  if (expected !== observed) out.push(`${path}: ${JSON.stringify(expected)} != ${JSON.stringify(observed)}`);
  return out;
}

// ---------------------------------------------------------------------------
// The run
// ---------------------------------------------------------------------------

if (!Array.isArray(suite.cases) || suite.cases.length === 0) {
  console.error('the conformance suite holds no cases — it would pass while checking nothing');
  process.exit(1);
}

let failures = 0;
const codesSeen = new Set();
for (const entry of suite.cases) {
  const { expect, ...caseSpec } = entry;
  const observed = observe(caseSpec);
  if (observed && observed.error && typeof observed.error.code === 'string') {
    codesSeen.add(observed.error.code);
  }
  const lines = compare(expect, observed, suite.tolerance);
  if (lines.length) {
    failures += 1;
    console.error(`FAIL ${caseSpec.id}`);
    for (const line of lines.slice(0, 8)) console.error(`       ${line}`);
    if (lines.length > 8) console.error(`       … and ${lines.length - 8} more`);
  }
}

// `Error::code()` is the stable half of the contract and this host is the one
// that actually exposes it (`throw()` sets `e.code`). So the browser runner is
// where "the codes are reachable through a binding" is really proved.
const missingCodes = suite.errorCodes.filter((c) => !codesSeen.has(c));
if (missingCodes.length) {
  failures += 1;
  console.error(`FAIL error codes never reached through this binding: ${missingCodes.join(', ')}`);
}

const label = `${suite.cases.length} cases, ${suite.errorCodes.length} error codes`;
if (failures) {
  console.error(`\nbrowser host (${pkgDir}): ${failures} conformance failure(s) over ${label}`);
  process.exit(1);
}
console.log(`browser host (${pkgDir}): conformance OK (${label})`);
